pytest_mh/_private/plugin.py

```diff
@@ -547,6 +570,16 @@ def _can_run_test(self, item: pytest.Item, data: MultihostItemData | None) -> bo
             if data.topology_mark.name not in self.mh_topology:
                 return False
 
+        if (
+            self._has_preferred_topology(item)
+            and self.mh_ignore_preferred_topology is False
+            and data.topology_mark is not None
+        ):
+            current_topology = data.topology_mark.name
+            preferred_topology = item.get_closest_marker(name="preferred_topology")
+            if preferred_topology:
+                return self._run_preferred_topology(preferred_topology, current_topology)
```

Comment:
This ignores the case where --mh-topology is set and `self._has_preferred_topology(item)` is actually checked twice (the second call is on line 580). You can rewrite it to:

```python
        # Run only for preferred topology unless specific topology is requested or the marker is ignored
        if not self.mh_topology and not self.mh_ignore_preferred_topology:
            preferred_topology = item.get_closest_marker(name="preferred_topology")
            if preferred_topology is not None and data.topology_mark is not None:
                return self._run_preferred_topology(preferred_topology, data.topology_mark.name)
```